presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsAnnotationProcessor.java

```diff
@@ -0,0 +1,247 @@
+        double averageRowSize = NaN;
+        double maxValue = scalarStatsHeader.getMax();
+        double minValue = scalarStatsHeader.getMin();
+        for (Map.Entry<Integer, ScalarPropagateSourceStats> paramIndexVsStatsMap : scalarStatsHeader.getArgumentStats().entrySet()) {
+            ScalarPropagateSourceStats scalarPropagateSourceStats = paramIndexVsStatsMap.getValue();
+            boolean propagateAllStats = scalarPropagateSourceStats.propagateAllStats();
+            nullFraction = min(firstFiniteValue(nullFraction, processSingleArgumentStatistic(
+                    sourceStats.stream().map(VariableStatsEstimate::getNullsFraction).collect(toImmutableList()),
+                    paramIndexVsStatsMap.getKey(),
+                    applyPropagateAllStats(propagateAllStats, scalarPropagateSourceStats.nullFraction()))), 1.0);
+            distinctValuesCount = firstFiniteValue(distinctValuesCount, processSingleArgumentStatistic(
+                    sourceStats.stream().map(VariableStatsEstimate::getDistinctValuesCount).collect(toImmutableList()), // an array list !
+                    paramIndexVsStatsMap.getKey(),
+                    applyPropagateAllStats(propagateAllStats, scalarPropagateSourceStats.distinctValuesCount())));
+            averageRowSize = firstFiniteValue(averageRowSize, processSingleArgumentStatistic(
+                    sourceStats.stream().map(VariableStatsEstimate::getAverageRowSize).collect(toImmutableList()),
+                    paramIndexVsStatsMap.getKey(),
+                    applyPropagateAllStats(propagateAllStats, scalarPropagateSourceStats.avgRowSize())));
+            maxValue = firstFiniteValue(maxValue, processSingleArgumentStatistic(
+                    sourceStats.stream().map(VariableStatsEstimate::getHighValue).collect(toImmutableList()),
+                    paramIndexVsStatsMap.getKey(),
+                    applyPropagateAllStats(propagateAllStats, scalarPropagateSourceStats.maxValue())));
+            minValue = firstFiniteValue(minValue, processSingleArgumentStatistic(
+                    sourceStats.stream().map(VariableStatsEstimate::getLowValue).collect(toImmutableList()),
+                    paramIndexVsStatsMap.getKey(),
+                    applyPropagateAllStats(propagateAllStats, scalarPropagateSourceStats.minValue())));
+        }
+        if (isNaN(maxValue) || isNaN(minValue)) {
+            minValue = NaN;
+            maxValue = NaN;
+        }
+        return VariableStatsEstimate.builder()
+                .setLowValue(minValue)
+                .setHighValue(maxValue)
+                .setNullsFraction(nullFraction)
+                .setAverageRowSize(minExcludingNaNs(firstFiniteValue(scalarStatsHeader.getAvgRowSize(), averageRowSize), getReturnTypeWidth(callExpression, UNKNOWN)))
+                .setDistinctValuesCount(processDistinctValuesCount(distinctValuesCount)).build();
+    }
+
+    private double getConstantDistinctValuesCount()
```

Comment:
Why is this a double instead of an integer? A number of distinct values feels fundamentally integral to me. If that's not what this is, then the method might need to be renamed.